scripts/SQLSERVER_Migration.sql

```diff
@@ -0,0 +1,46 @@
+DROP TABLE IF EXISTS [dbo].[TagTable];
+
+DROP FUNCTION IF EXISTS [dbo].[Split];
+
+CREATE TABLE dbo.TagTable(
+  ordering_id BIGINT NOT NULL,
+  tag NVARCHAR(64) NOT NULL,
+  PRIMARY KEY  (ordering_id, tag)
+);
+GO
+
+CREATE FUNCTION [dbo].[Split](@String varchar(8000), @Delimiter char(1))
+	RETURNS @temptable TABLE (items varchar(8000)) AS
+	BEGIN
+		DECLARE @idx INT
+		DECLARE @slice VARCHAR(8000)
+
+		SELECT @idx = 1
+		IF LEN(@String) < 1 OR @String is NULL
+			RETURN
+
+		WHILE @idx != 0 
+		BEGIN
+			SET @idx = CHARINDEX(@Delimiter, @String)
+			IF @idx != 0
+				SET @slice = LEFT(@String,@idx - 1)
+			ELSE
+				SET @slice = @String
+
+			IF(LEN(@slice) > 0)
+				INSERT INTO @temptable(Items) VALUES(@slice)     
+
+			SET @String = RIGHT(@String, LEN(@String) - @idx)
+			IF len(@String) = 0 
+				BREAK
+		END
+		RETURN
+	END;
+GO
+
+BEGIN TRANSACTION 
+INSERT INTO dbo.TagTable(ordering_id, tag)
```

Comment:
The other way we could go about making these scripts safe to run multiple times is rather than delete / recreate we can try the following idempotent pattern:

- create if not exists (for tables, functions)
- for each item to be added to the tag table, do an upsert (insert if not exist, update row if does exist)

That way running the script multiple times would be safe.

Reply:
Actually instead of doing an upsert, we could just skip the insert since the data is immutable -therefore no change.

Reply:
IMO these should be grabbing batches of items rather than trying to do the whole table. Otherwise for large tables this could run for some unknown time before erroring out with a transaction size issue. (Also in general batching will make it easier to run 'on the fly' without locking too much, especially for something like MSSQL.)

Reply:
Forgive how rusty my SQL experience is here, but you're talking about doing pagination with cursors right?

Reply:
Not cursors... the overall flow I was imagining (at least, for 0 downtime migration) was along the lines of

1. deploy everything in 'hybrid' mode (where we write both old style tags and new tags)
2. Once everything is deployed in that mode, run a script in the background, psuedo-sql:

```
declare @moveUpTo = Select max(ordering) from journal_table 
where (tags is not null and strlen(tags)>0)
and ordering not in (select ordering from journal_tag_table);

declare @startAtOrdering = 0;

while (@startAtOrdering < @moveUpTo)
begin
     insert into journal tag table (select tags from journal_table where orderingId >= @startAtOrdering && orderingId < @startAtOrdering+1000); --Obviously need more transform here
     @startAtOrdering = @startAtOrdering+1000;
end

```

you might need to think about how you do the transactions here... but hopefully gets the point across.

Reply:
> deploy everything in 'hybrid' mode (where we write both old style tags and new tags)

That makes sense and I think this can be done easily since we're not getting rid of the old tag column - does that make sense to you @Arkatufus ?

Reply:
> Once everything is deployed in that mode, run a script in the background, psuedo-sql:

I see what you mean: breaking up the inserts into transactions with groups of N "tag rows" at a time. I think that should be doable.